### PR TITLE
Timeouts support for GStreamer backend

### PR DESCRIPTION
**Pull request by asmorkalov**
Address https://github.com/opencv/opencv/issues/22868
Used the same defaults as it's done for FFmpeg

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

```
force_builders=Custom
build_image:Custom=gstreamer:16.04
buildworker:Custom=linux-1
```

### CODE REVIEW

**Comment by asmorkalov** [comment]
@alalek Could you take a look?

**Comment by mshabunin** [comment]
@asmorkalov , I've added the test to your PR. Also made some cosmetic changes to reduce number of constant being used. And added timeout to the `startPipeline` method, but I'm not sure about this. Please take a look.

**Comment by asmorkalov** [comment]
@mshabunin Thanks a lot for the fix!

File: modules/videoio/src/cap_gstreamer.cpp [hunk 1]
```diff
@@ -504,7 +510,7 @@ bool GStreamerCapture::grabFrame()
     if (gst_app_sink_is_eos(GST_APP_SINK(sink.get())))
         return false;
 
-    sample.attach(gst_app_sink_pull_sample(GST_APP_SINK(sink.get())));
+    sample.attach(gst_app_sink_try_pull_sample(GST_APP_SINK(sink.get()), readTimeout));
```
[the remark this answers is not in the document]
Comment (asmorkalov) [reply modules/videoio/src/cap_gstreamer.cpp:513]: I have device with old GSteamer. Will check today and update.

File: modules/videoio/test/test_gstreamer.cpp [hunk 2]
```diff
@@ -151,4 +151,22 @@ TEST(videoio_gstreamer, gray16_writing)
...
+    if (!videoio_registry::hasBackend(CAP_GSTREAMER))
+        throw SkipTestException("GStreamer backend was not found");
+    VideoCapture cap;
+    cap.open("videotestsrc ! appsink", CAP_GSTREAMER);
+    ASSERT_TRUE(cap.isOpened());
+    const double default_timeout = 30000; // 30 seconds
+    const double open_timeout = 5678; // 3 seconds
+    const double read_timeout = 1234; // 1 second
+    EXPECT_NEAR(default_timeout, cap.get(CAP_PROP_OPEN_TIMEOUT_MSEC), 1e-3);
+    EXPECT_NEAR(default_timeout, cap.get(CAP_PROP_READ_TIMEOUT_MSEC), 1e-3);
```
Comment (alalek) [inline modules/videoio/test/test_gstreamer.cpp:165]: If `cap.get(CAP_PROP_READ_TIMEOUT_MSEC)` is zero then we should bailout or skip this tests (as GStreamer version is very old).

Currently it fails:

```
[ RUN      ] videoio_gstreamer.timeout_property
[ WARN:0@10.679] global cap_gstreamer.cpp:1427 open OpenCV | GStreamer warning: Cannot query video position: status=1, value=0, duration=-1
[ INFO:0@10.679] global cap_gstreamer.cpp:1126 open OpenCV | GStreamer: videotestsrc ! appsink
[ INFO:0@10.679] global cap_gstreamer.cpp:1159 open OpenCV | GStreamer: mode - MANUAL
/build/precommit_custom_linux/4.x/opencv/modules/videoio/test/test_gstreamer.cpp:165: Failure
The difference between default_timeout and cap.get(CAP_PROP_READ_TIMEOUT_MSEC) is 30000, which exceeds 1e-3, where
default_timeout evaluates to 30000,
cap.get(CAP_PROP_READ_TIMEOUT_MSEC) evaluates to 0, and
1e-3 evaluates to 0.001.
[ WARN:0@10.680] global cap_gstreamer.cpp:1772 setProperty OpenCV | GStreamer warning: GStreamer before 1.10 does not support read timeout
/build/precommit_custom_linux/4.x/opencv/modules/videoio/test/test_gstreamer.cpp:169: Failure
The difference between read_timeout and cap.get(CAP_PROP_READ_TIMEOUT_MSEC) is 1234, which exceeds 1e-3, where
read_timeout evaluates to 1234,
cap.get(CAP_PROP_READ_TIMEOUT_MSEC) evaluates to 0, and
1e-3 evaluates to 0.001.
[  FAILED  ] videoio_gstreamer.timeout_property (2 ms)
```

**Review by mshabunin — APPROVED** [review]

**Comment by canberkgurel** [comment]
@asmorkalov when a read timeout occurs, how can I detect a timeout and handle it gracefully? 

Also, what's the expected behavior of `cap.read(img);`? I ran a test and I can confirm that `cap.read` didn't return `false` when I killed the Gstreamer transmitter.

**Pull request merged**